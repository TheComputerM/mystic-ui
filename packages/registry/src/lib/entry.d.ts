interface RegistryEntry {
	name: string;
	npmDeps?: string[];
}
