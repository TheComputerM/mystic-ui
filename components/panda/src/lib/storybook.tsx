import { lazy } from "solid-js";

const sources = import.meta.glob("../stories/**/*.tsx", {
	eager: true,
	query: "?raw",
	import: "default",
});

export function getStory(name: string, example: string) {
	const Component = lazy(() => import(`../stories/${name}/${example}.tsx`));
	return {
		// biome-ignore lint/suspicious/noExplicitAny: props are different for each component
		render: (props: any) => <Component {...props} />,
		parameters: {
			docs: {
				source: {
					code: sources[`../stories/${name}/${example}.tsx`],
				},
			},
		},
	};
}
