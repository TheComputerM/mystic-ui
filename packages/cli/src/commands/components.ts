import type { RegistryEntry } from "@mystic-ui/registry/src/schema";

async function getRegistryEntry(component: string) {
	const res = await fetch(
		`https://raw.githubusercontent.com/TheComputerM/mystic-ui/main/packages/registry/build/${component}.json`,
	);
	const data: RegistryEntry = await res.json();
	return data;
}

export async function addComponent() {}
