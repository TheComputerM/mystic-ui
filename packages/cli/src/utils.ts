import { CommanderError } from "commander";
import { cosmiconfig } from "cosmiconfig";
import { z } from "zod";

export const FRAMEWORKS = ["panda", "tailwind"] as const;

const baseSchema = z.object({ outputPath: z.string(), configPath: z.string() });

export const configSchema = z.discriminatedUnion("framework", [
	z.object({ framework: z.literal("panda") }).merge(baseSchema),
	z
		.object({
			framework: z.literal("tailwind"),
			tailwind: z.object({ aliases: z.object({ utils: z.string() }) }),
		})
		.merge(baseSchema),
]);

// export const configSchema = z
// 	.object({
// 		framework: z.enum(FRAMEWORKS),
// 		outputPath: z.string(),
// 		configPath: z.string(),
// 		tailwind: z
// 			.object({
// 				aliases: z.object({
// 					utils: z.string(),
// 				}),
// 			})
// 			.optional(),
// 	})
// 	.superRefine((val, ctx) => {
// 		// If the framework is tailwind, the tailwind config is required
// 		if (val.framework === "tailwind" && !val.tailwind) {
// 			ctx.addIssue({
// 				code: "custom",
// 				message: "Additional config is required for tailwind framework",
// 				path: ["tailwind"],
// 			});
// 		}
// 	});

export type ConfigSchema = z.infer<typeof configSchema>;

/**
 * Get the mystic config from the project
 */
export async function getMysticConfig() {
	const explorer = cosmiconfig("mystic", {
		searchPlaces: ["mystic.config.json", ".config/mystic.config.json"],
		searchStrategy: "project",
	});
	const result = await explorer.search();

	if (!result) {
		throw new CommanderError(
			1,
			"no-config",
			"No mystic.config.json found, generate one using the init command.",
		);
	}

	const parsed = configSchema.safeParse(result.config);

	if (!parsed.success) {
		throw new CommanderError(
			1,
			"invalid-config",
			`${parsed.error.message}\nYour mystic.config.json is outdated/invalid, please update it.`,
		);
	}

	return { config: parsed.data, filepath: result.filepath };
}
