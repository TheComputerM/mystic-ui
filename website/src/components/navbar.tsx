import { createListCollection } from "@ark-ui/solid";
import { TbBrandGithubFilled, TbChevronDown } from "solid-icons/tb";
import { Index } from "solid-js";
import { css } from "styled-system/css";
import { Box, Container, Divider, HStack } from "styled-system/jsx";
import { ThemeSwitcher } from "./theme-switcher";
import { Badge } from "./ui/badge";
import { IconButton } from "./ui/icon-button";
import { Select } from "./ui/select";

const SelectCSSFramework = () => {
	interface Item {
		label: string;
		value: string;
		disabled?: boolean;
	}

	const collection = createListCollection<Item>({
		items: [
			{ label: "Tailwind", value: "tailwind" },
			{ label: "Panda", value: "panda" },
			{ label: "UnoCSS", value: "unocss", disabled: true },
		],
	});

	return (
		<Select.Root
			collection={collection}
			variant="ghost"
			defaultValue={["tailwind"]}
			size="sm"
		>
			<Select.Control>
				<Select.Trigger>
					<Select.ValueText fontWeight="medium" />
					<TbChevronDown />
				</Select.Trigger>
			</Select.Control>
			<Select.Positioner>
				<Select.Content>
					<Index each={collection.items}>
						{(item) => (
							<Select.Item item={item()}>
								<Select.ItemText>{item().label}</Select.ItemText>
							</Select.Item>
						)}
					</Index>
				</Select.Content>
			</Select.Positioner>
		</Select.Root>
	);
};

export const Navbar = () => {
	return (
		<Box py="2" position="sticky" top="0" zIndex="docked">
			<Container>
				<HStack justify="space-between">
					<HStack>
						<a href="/" class={css({ textStyle: "lg", fontWeight: "bold" })}>
							Mystic UI
						</a>
						<Badge>alpha</Badge>
					</HStack>
					<HStack gap="1">
						<SelectCSSFramework />
						<Divider h="6" mx="2" orientation="vertical" />
						<IconButton
							size="sm"
							variant="ghost"
							asChild={(parentProps) => (
								<a
									href="https://github.com/TheComputerM/mystic-ui"
									target="_blank"
									{...parentProps()}
								>
									<TbBrandGithubFilled />
								</a>
							)}
						/>
						<ThemeSwitcher />
					</HStack>
				</HStack>
			</Container>
		</Box>
	);
};
