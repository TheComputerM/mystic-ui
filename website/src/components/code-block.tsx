import { useClipboard } from "@ark-ui/solid";
import { createAsync } from "@solidjs/router";
import type { BundledLanguage } from "shiki";
import { TbCheck, TbCopy } from "solid-icons/tb";
import type { Component } from "solid-js";
import { css } from "styled-system/css";
import { Box } from "styled-system/jsx";
import { highlight } from "~/lib/shiki";
import { Clipboard } from "./ui/clipboard";
import { IconButton } from "./ui/icon-button";

interface CodeBlockProps {
	code: string;
	lang?: BundledLanguage;
}

export const RawCodeBlock: Component<CodeBlockProps & { html?: string }> = (
	props,
) => {
	const clipboard = useClipboard({ value: props.code });

	return (
		<Clipboard.RootProvider value={clipboard} position="relative" class="dark">
			<Box
				textStyle="sm"
				class={css({
					"&>pre": {
						padding: 4,
						borderWidth: "1px",
						borderColor: "border.accent",
						borderRadius: "l3",
						tabSize: 2,
						overflowX: "auto",
					},
				})}
				innerHTML={props.html}
			/>
			<Clipboard.Control position="absolute" right="2" top="2">
				<Clipboard.Trigger
					asChild={(parentProps) => (
						<IconButton size="sm" variant="ghost" {...parentProps()}>
							<Clipboard.Indicator copied={<TbCheck />}>
								<TbCopy />
							</Clipboard.Indicator>
						</IconButton>
					)}
				/>
			</Clipboard.Control>
		</Clipboard.RootProvider>
	);
};

export const CodeBlock: Component<CodeBlockProps> = (props) => {
	const html = createAsync(() => highlight(props.code, props.lang));
	return <RawCodeBlock html={html()} code={props.code} />;
};
