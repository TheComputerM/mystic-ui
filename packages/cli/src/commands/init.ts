import fs from "node:fs";
import path from "node:path";
import * as p from "@clack/prompts";
import { cosmiconfig } from "cosmiconfig";
import { FRAMEWORKS, configSchema, getMysticConfig } from "../utils";

/**
 * Detects the CSS framework used in the project
 */
async function detectFramework() {
	const configs = await Promise.all(
		FRAMEWORKS.map((framework) => ({
			framework,
			explorer: cosmiconfig(framework, { searchStrategy: "project" }),
		})).map(async ({ framework, explorer }) => ({
			framework,
			result: await explorer.search(),
		})),
	);
	return configs.find((config) => config.result !== null);
}

/**
 * Wrapper around p.group that exits the process on cancel
 */
async function promptGroup<T>(prompts: p.PromptGroup<T>) {
	return p.group(prompts, {
		onCancel: () => {
			p.cancel("Operation cancelled.");
			process.exit(0);
		},
	});
}

/**
 * Creates a mystic.config.json file by asking prompts
 */
async function createConfig() {
	p.intro("Welcome to Mystic UI");

	const resolvedConfig = await detectFramework();

	const config = await promptGroup({
		framework: () =>
			p.select({
				message: "Which CSS framework are you using?",
				options: FRAMEWORKS.map((value) => ({ label: value, value })),
				initialValue: resolvedConfig?.framework,
			}),
		configPath: ({ results }) =>
			p.text({
				message: "Where is your CSS framework config?",
				initialValue: resolvedConfig?.result
					? path.relative(process.cwd(), resolvedConfig.result.filepath)
					: `${results.framework}.config.ts`,
				validate(value) {
					if (!value) return "Please enter a path.";
					if (!fs.existsSync(path.join(process.cwd(), value)))
						return "File doesn't exist. Init your CSS framework config first.";
				},
			}),
		outputPath: () =>
			p.text({
				message: "Where would you like to store your components?",
				initialValue: "./src/components/mystic-ui/",
				validate(value) {
					if (!value) return "Please enter a path.";
					if (!value.startsWith("."))
						return "Please enter a relative path to the project root.";
				},
			}),
		tailwind: ({ results }) => {
			if (results.framework !== "tailwind") {
				return undefined;
			}
			return promptGroup({
				aliases: () =>
					promptGroup({
						utils: () =>
							p.text({
								message: "Import alias for where the cn utility is located",
								initialValue: "@/lib/utils",
								validate(value) {
									if (!value) return "Please enter a path.";
								},
							}),
					}),
			});
		},
	});

	configSchema.parse(config);

	fs.writeFileSync("mystic.config.json", JSON.stringify(config, null, 2));

	p.outro("You're all set!");
}

export default async function initCommand() {
	try {
		// check if a mystic.config.json already exists
		await getMysticConfig();
		console.log(
			"A valid config is already present. Delete it if you want to regenerate the config.",
		);
	} catch {
		// create a new config file
		await createConfig();
	}
}
