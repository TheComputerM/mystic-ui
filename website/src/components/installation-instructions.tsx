import type { RegistryEntry } from "@mystic-ui/registry/src/schema";
import { createAsync, useParams } from "@solidjs/router";
import { type Component, For, Show } from "solid-js";
import { getRegistryEntry } from "~/lib/registry";
import { CodeBlock } from "./code-block";
import { Step, Steps } from "./ui/stepper";
import { Tabs } from "./ui/tabs";

const instructions: {
	title: string;
	condition: (entry: RegistryEntry) => boolean;
	component: Component<{ entry: RegistryEntry }>;
}[] = [
	{
		title: "Install dependencies",
		condition: (entry) => !!entry.dependencies,
		component: (props) => (
			<CodeBlock
				code={`npm i ${props.entry.dependencies?.join(" ")}`}
				lang="shell"
			/>
		),
	},
	{
		title: "Copy and paste the component.",
		condition: () => true,
		component: (props) => {
			const params = useParams() satisfies {
				framework: "tailwind" | "panda";
			};
			return <CodeBlock code={props.entry[params.framework].file} lang="tsx" />;
		},
	},
];

export const InstallationInstructions: Component<{ component: string }> = (
	props,
) => {
	const entry = createAsync(() => getRegistryEntry(props.component));

	return (
		<Tabs.Root defaultValue="cli">
			<Tabs.List>
				<Tabs.Trigger value="cli">CLI</Tabs.Trigger>
				<Tabs.Trigger value="manual">Manual</Tabs.Trigger>
				<Tabs.Indicator />
			</Tabs.List>
			<Tabs.Content value="cli">
				<CodeBlock code={`npx mystic-ui add ${props.component}`} lang="shell" />
			</Tabs.Content>
			<Tabs.Content value="manual">
				<Steps>
					<Show when={entry()}>
						<For
							each={instructions.filter(({ condition }) => condition(entry()!))}
						>
							{(instruction, i) => (
								<Step number={i() + 1} title={instruction.title}>
									<instruction.component entry={entry()!} />
								</Step>
							)}
						</For>
					</Show>
				</Steps>
			</Tabs.Content>
		</Tabs.Root>
	);
};
