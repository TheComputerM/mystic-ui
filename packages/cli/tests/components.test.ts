import { describe, expect, test } from "bun:test";
import { generateCode, parseModule } from "magicast";
import { mergeConfig, transform } from "../src/commands/components";

describe("transform()", () => {
	test("should fix imports", async () => {
		const output = transform(
			`
import { cn } from "@/lib/utils";
import test from "some-module";
      `,
			{
				framework: "tailwind",
				configPath: "tailwind.config.mjs",
				outputPath: "./src/components/mystic-ui/",
				tailwind: {
					aliases: {
						utils: "mystic-ui/lib/utils",
					},
				},
			},
		);

		expect(output).toBe(`
import { cn } from "mystic-ui/lib/utils";
import test from "some-module";
      `);
	});
});

describe("mergeConfig()", () => {
	test("tailwind.config.js", () => {
		const mod = parseModule(`/** @type {import('tailwindcss').Config} */
export default {
  content: [],
  theme: {
    extend: {},
  },
  plugins: [],
}`);
		mergeConfig(mod, {
			theme: {
				extend: {
					animation: {
						grid: "grid 15s linear infinite",
					},
					keyframes: {
						grid: {
							"0%": { transform: "translateY(-50%)" },
							"100%": { transform: "translateY(0)" },
						},
					},
				},
			},
		});

		const { code } = generateCode(mod);
		expect(code).toMatchSnapshot("tailwind.config.js");
	});

	test("panda.config.ts", () => {
		const mod = parseModule(`import { defineConfig } from "@pandacss/dev";

export default defineConfig({
	// Whether to use css reset
	preflight: true,

	presets: ["@pandacss/preset-panda"],

	// Where to look for your css declarations
	include: ["./src/**/*.{js,jsx,ts,tsx}", "./pages/**/*.{js,jsx,ts,tsx}"],

	// Files to exclude
	exclude: [],

	// Useful for theme customization
	theme: {},

	// The output directory for your css system
	outdir: "styled-system",

	jsxFramework: "solid",
});
`);
		mergeConfig(mod, {
			theme: {
				extend: {
					tokens: {
						animations: {
							grid: {
								value: "grid 15s linear infinite",
							},
						},
					},
					keyframes: {
						grid: {
							"0%": { transform: "translateY(-50%)" },
							"100%": { transform: "translateY(0)" },
						},
					},
				},
			},
		});

		const { code } = generateCode(mod);
		expect(code).toMatchSnapshot("panda.config.ts");
	});
});
