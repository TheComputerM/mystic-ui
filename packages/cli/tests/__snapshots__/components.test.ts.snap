// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`mergeConfig() tailwind.config.js: tailwind.config.js 1`] = `
"/** @type {import('tailwindcss').Config} */
export default {
  content: [],
  theme: {
    extend: {
      animation: {
        grid: 'grid 15s linear infinite',
      },

      keyframes: {
        grid: {
          '0%': {
            transform: 'translateY(-50%)',
          },

          '100%': {
            transform: 'translateY(0)',
          },
        },
      },
    },
  },
  plugins: [],
}"
`;

exports[`mergeConfig() panda.config.ts: panda.config.ts 1`] = `
"import { defineConfig } from "@pandacss/dev";

export default defineConfig({
				// Whether to use css reset
				preflight: true,

				presets: ["@pandacss/preset-panda"],

				// Where to look for your css declarations
				include: ["./src/**/*.{js,jsx,ts,tsx}", "./pages/**/*.{js,jsx,ts,tsx}"],

				// Files to exclude
				exclude: [],

				// Useful for theme customization
				theme: {
				 extend: {
					 tokens: {
						 animations: {
							 grid: {
								 value: "grid 15s linear infinite",
								},
							},
						},

					 keyframes: {
						 grid: {
							 "0%": {
								 transform: "translateY(-50%)",
								},

							 "100%": {
								 transform: "translateY(0)",
								},
							},
						},
					},
				},

				// The output directory for your css system
				outdir: "styled-system",

				jsxFramework: "solid",
});"
`;
